controllers/apply-next/helpers.js

```diff
@@ -0,0 +1,186 @@
+'use strict';
+const { cloneDeep, find, findIndex, flatMap, isEmpty, pick } = require('lodash');
+const { get, getOr } = require('lodash/fp');
+
+const FORM_STATES = {
+    incomplete: {
+        type: 'incomplete',
+        label: 'In progress'
+    },
+    complete: {
+        type: 'complete',
+        label: 'Complete'
+    },
+    invalid: {
+        type: 'invalid',
+        label: 'Invalid'
+    },
+    empty: {
+        type: 'empty',
+        label: 'Not started'
+    }
+};
+
+function prepareForm(locale, formModel, formData) {
+    const localise = get(locale);
+
+    const translateSection = section => {
+        section.title = localise(section.title);
+        if (section.introduction) {
+            section.introduction = localise(section.introduction);
+        }
+        return section;
+    };
+
+    const translateField = field => {
+        field.label = localise(field.label);
+        field.explanation = localise(field.explanation);
+        const match = find(formData, (value, name) => name === field.name);
+        if (match) {
+            field.value = match;
+        }
+
+        // Translate each option (if set)
+        if (field.options) {
+            field.options = field.options.map(option => {
+                option.label = localise(option.label);
+                option.explanation = localise(option.explanation);
+                return option;
+            });
+        }
+        return field;
+    };
+
+    const translateStep = step => {
+        step.title = localise(step.title);
+
+        step.fieldsets = step.fieldsets.map(fieldset => {
+            fieldset.legend = localise(fieldset.legend);
+            fieldset.introduction = localise(fieldset.introduction);
+            fieldset.fields = fieldset.fields.map(translateField);
+            return fieldset;
+        });
+
+        return step;
+    };
+
+    const clonedForm = cloneDeep(formModel);
+    clonedForm.title = localise(formModel.title);
+
+    clonedForm.sections = clonedForm.sections.map(section => {
+        section = translateSection(section);
+        section.steps = section.steps.map(step => translateStep(step));
```

Comment:
This can just be:

`section.steps = section.steps.map(translateStep);`

Reply:
Personally I prefer the explicitness. `section.steps.map(translateStep)` implicitly passes `translateStep(currentValue, index, array)` where the long-hand only passes the explicit arguments.